File: itam-system/app/schemas/inventory.py
from app.core.time import TimezoneModel

from datetime import datetime
from typing import Optional

from pydantic import ConfigDict, Field


class InventoryItemBase(TimezoneModel):
    item_type: str
    code: str
    name: str
    brand: Optional[str] = None
    model: Optional[str] = None
    spec: Optional[str] = None
    total_qty: int = 0
    available_qty: Optional[int] = None
    min_qty: int = 0
    unit_cost: float = 0
    license_key: Optional[str] = None
    expire_date: Optional[datetime] = None
    supplier: Optional[str] = None
    dept_id: Optional[str] = None
    location: Optional[str] = None
    status: str = "active"
    remark: Optional[str] = None


class InventoryItemCreate(InventoryItemBase):
    pass


class InventoryItemUpdate(TimezoneModel):
    item_type: Optional[str] = None
    code: Optional[str] = None
    name: Optional[str] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    spec: Optional[str] = None
    total_qty: Optional[int] = None
    available_qty: Optional[int] = None
    min_qty: Optional[int] = None
    unit_cost: Optional[float] = None
    license_key: Optional[str] = None
    expire_date: Optional[datetime] = None
    supplier: Optional[str] = None
    dept_id: Optional[str] = None
    location: Optional[str] = None
    status: Optional[str] = None
    remark: Optional[str] = None


class InventoryItemOut(InventoryItemBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    available_qty: int
    assigned_qty: int
    created_at: datetime
    updated_at: datetime


class InventoryLedgerCreate(TimezoneModel):
    action: str
    quantity: int = 1
    assignee_user_id: Optional[str] = None
    assignee_name: Optional[str] = None
    dept_id: Optional[str] = None
    asset_id: Optional[str] = None
    location: Optional[str] = None
    remark: Optional[str] = None


class InventoryLedgerOut(InventoryLedgerCreate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    item_id: int
    operator: str
    created_at: datetime


class LicenseSeatBatchCreate(TimezoneModel):
    count: int = Field(default=1, ge=1, le=1000)
    seat_codes: list[str] = Field(default_factory=list)
    remark: Optional[str] = None


class LicenseSeatAssign(TimezoneModel):
    assignee_user_id: Optional[str] = None
    assignee_name: Optional[str] = None
    dept_id: Optional[str] = None
    asset_id: Optional[str] = None
    remark: Optional[str] = None


class LicenseSeatReturn(TimezoneModel):
    remark: Optional[str] = None


class LicenseSeatOut(TimezoneModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    item_id: int
    seat_code: str
    status: str
    assignee_user_id: Optional[str] = None
    assignee_name: Optional[str] = None
    dept_id: Optional[str] = None
    asset_id: Optional[str] = None
    assigned_at: Optional[datetime] = None
    returned_at: Optional[datetime] = None
    remark: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class LicenseSeatHistoryOut(TimezoneModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    seat_id: int
    action: str
    assignee_user_id: Optional[str] = None
    assignee_name: Optional[str] = None
    dept_id: Optional[str] = None
    asset_id: Optional[str] = None
    operator: str
    remark: Optional[str] = None
    created_at: datetime


class LicenseSeatPage(TimezoneModel):
    list: list[LicenseSeatOut]
    total: int
    page: int
    page_size: int
    summary: dict[str, int]


class ComponentInstallationOut(TimezoneModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    item_id: int
    asset_id: str
    asset_name: Optional[str] = None
    quantity: int
    dept_id: Optional[str] = None
    installed_by: str
    installed_at: datetime
    remark: Optional[str] = None
    updated_at: datetime


class ComponentInstallationPage(TimezoneModel):
    list: list[ComponentInstallationOut]
    total: int
    page: int
    page_size: int

File: itam-frontend/src/api/inventory.js
import request from '../utils/request'

export const inventoryTypes = [
  { label: '许可证', value: 'license' },
  { label: '耗材', value: 'consumable' },
  { label: '配件', value: 'accessory' },
  { label: '组件', value: 'component' }
]

export const inventoryActions = [
  { label: '入库', value: 'in' },
  { label: '出库调整', value: 'out' },
  { label: '分配给人员', value: 'assign' },
  { label: '耗用', value: 'consume' },
  { label: '归还', value: 'return' },
  { label: '装配到资产', value: 'install' },
  { label: '从资产拆卸', value: 'uninstall' },
  { label: '增加调整', value: 'adjust_add' },
  { label: '减少调整', value: 'adjust_sub' }
]

export async function getInventoryItems(params = {}) {
  const result = await request.get('/inventory/items', {
    params: {
      keyword: params.keyword || undefined,
      item_type: params.item_type || undefined,
      item_types: params.item_types || undefined,
      status: params.status || undefined,
      low_stock: params.low_stock || undefined,
      expiring_days: params.expiring_days ?? undefined,
      page: params.page || undefined,
      page_size: params.page_size ?? params.pageSize ?? undefined
    }
  })
  return {
    list: (result.list || []).map(mapInventoryItem),
    total: Number(result.total || 0),
    summary: result.summary || {}
  }
}

export function createInventoryItem(payload) {
  return request.post('/inventory/items', normalizeItem(payload))
}

export function updateInventoryItem(id, payload) {
  return request.put(`/inventory/items/${id}`, normalizeItem(payload))
}

export function operateInventoryItem(id, payload) {
  return request.post(`/inventory/items/${id}/ledger`, {
    action: payload.action,
    quantity: Number(payload.quantity || 1),
    assignee_user_id: payload.assignee_user_id || '',
    assignee_name: payload.assignee_name || '',
    dept_id: payload.dept_id || '',
    asset_id: payload.asset_id || '',
    location: payload.location || '',
    remark: payload.remark || ''
  })
}

export async function getInventoryLedger(id) {
  const rows = await request.get(`/inventory/items/${id}/ledger`)
  return (rows || []).map(row => ({
    ...row,
    action_label: actionLabel(row.action),
    created_at_text: formatDateTime(row.created_at)
  }))
}

export function typeLabel(value) {
  return inventoryTypes.find(item => item.value === value)?.label || value || '-'
}

export function actionLabel(value) {
  return inventoryActions.find(item => item.value === value)?.label || value || '-'
}

function normalizeItem(payload) {
  return {
    item_type: payload.item_type,
    code: payload.code,
    name: payload.name,
    brand: payload.brand || '',
    model: payload.model || '',
    spec: payload.spec || '',
    total_qty: Number(payload.total_qty || 0),
    available_qty: payload.available_qty === '' || payload.available_qty == null ? undefined : Number(payload.available_qty),
    min_qty: Number(payload.min_qty || 0),
    unit_cost: Number(payload.unit_cost || 0),
    license_key: payload.license_key || '',
    expire_date: payload.expire_date ? `${payload.expire_date}T00:00:00` : null,
    supplier: payload.supplier || '',
    dept_id: payload.dept_id || '',
    location: payload.location || '',
    status: payload.status || 'active',
    remark: payload.remark || ''
  }
}

function mapInventoryItem(row) {
  return {
    ...row,
    type_label: typeLabel(row.item_type),
    expire_date_text: formatDate(row.expire_date),
    low_stock: Number(row.available_qty || 0) <= Number(row.min_qty || 0)
  }
}

function formatDate(value) {
  if (!value) return ''
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) return ''
  return date.toISOString().slice(0, 10)
}

function formatDateTime(value) {
  if (!value) return ''
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) return ''
  return date.toLocaleString('zh-CN', { hour12: false })
}
